Fix inductive implicits performance regression

milessabin opened the pull request:
The fix for #6058 in #6163 caused a significant performance regression in the inductive implicits benchmark because the use of `=:=` rather than `==` in the divergence check was significantly slower. It is the right test however, so we need a quicker check to rule out negative cases.

We're already computing the covering sets and sizes of the two types being compared. `tp1 =:= tp2` should entail `(sz1 == sz2 && cs1 == cs2)`, so the contrapositive `(sz1 != sz2 || cs1 != cs2)` should entail that `!(tp1 =:= tp2)`. However the covering set and size computations were incorrect (they missed types mentioned in bounds which should have been included, and included symbols for unsolved type variables which should not).

This commit fixes the latter issue, which allows covering set and size tests to be used to avoid expensive full type equality tests.

Fixes the inductive implicits part of #6300.

milessabin commented:
test performance please

milessabin commented:
test performance please

smarter commented:
@milessabin Looks like you're not in the whitelist: https://github.com/liufengyun/bench/blob/f8f08f6c31f561f9b831c02b3281609a46650417/bin/process#L22, you can make a PR to fix that.

odersky commented:
test performance please

milessabin commented:
@smarter done here: https://github.com/liufengyun/bench/pull/723

liufengyun/bench seems to be fresher than lampepfl/bench ... how come?

milessabin commented:
That seems to have done the trick :-)

liufengyun commented:
Just add @milessabin and @anatoliykmetyuk . Hot maintenance happens in liufengyun/bench , so it is fresher.

odersky commented:
test performance please

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -4973,6 +4973,14 @@ object Types {
+      case tp: TermRef =>
+        apply(n, tp.underlying)
+      case tp: TypeParamRef =>
+        ctx.typerState.constraint.entry(tp) match {
```
I believe this is equivalent to
```
apply(n, ctx.typeComparer.bounds(tp))
```
But that also points to a problem: What if the type is F-bounded? Would that not give an infinite recursion?

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -4973,6 +4973,14 @@ object Types {
+        apply(n, tp.underlying)
+      case tp: TypeParamRef =>
+        ctx.typerState.constraint.entry(tp) match {
+          case tb: TypeBounds => foldOver(n, tb)
```
This case seems to be redundant. Is this not covered by the third case?
Also, what if the type is F-bounded? Would that not give an infinite recursion?

milessabin replied:
The type being F-bounded _is_ a problem. But we do need to look into those bounds. What are the options here?

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -4990,10 +4998,18 @@ object Types {
-        case tp: TypeBounds =>
-          foldOver(cs, tp)
-        case other =>
+        case tp: TypeRef if tp.prefix.isValueType =>
```
What does the `prefix.isValueType` condition accomplish?

milessabin replied:
It was a half-baked attempt at `tp.typeSymbol.isClass`.

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -4990,10 +4998,18 @@ object Types {
+        case tp: TermRef =>
+          apply(cs, tp.underlying)
+        case tp: TypeParamRef =>
+          ctx.typerState.constraint.entry(tp) match {
```
Same argument as before. First case looks redundant, and what about F-bounds?

milessabin commented:
@odersky I've eliminated the redundant cases and switched to `apply(n, ctx.typeComparer.bounds(tp))` as you suggested. I've added a `seen` set to handle with F-bounds and a test which exercise it (it SOEs, without the seen set).

milessabin commented:
test performance please

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -4966,41 +4966,57 @@ object Types {
-        apply(n, ctx.typeComparer.bounds(tp))
-      case _ =>
-        foldOver(n, tp)
+    val seen: util.HashSet[Type] = new util.HashSet[Type](64) {
```
It's probably faster to use a `java.util.IdentityHashMap[Type, Type]`, We already use that in lots of other places in the compiler.

allanrenucci replied:
Or even:
```scala
import java.util.{Collections, IdentityHashMap}
val seen = Collections.newSetFromMap[Type](new IdentityHashMap)
```

milessabin commented:
Now with `IdentityHashMap`. Will merge when green.

milessabin commented:
test performance please

milessabin commented:
The CI failures look to be timeout and memory related, and I can't reproduce locally, so I'll restart the CI job and hope for the best.

milessabin commented:
That fixed it. Merging now.
